runtime-modules/project-token/src/lib.rs

```diff
@@ -237,40 +234,209 @@ decl_module! {
+                Self::ensure_account_data_exists(token_id, &sender)?;
+            }
+
+            // Ensure vesting schedule can added if doesn't already exist
+            // (MaxVestingSchedulesPerAccountPerToken not exceeded)
+            let acc_data = AccountInfoByTokenAndAccount::<T>::get(token_id, &sender);
+            let vesting_cleanup_key = acc_data.ensure_can_add_or_update_vesting_schedule::<T>(
+                current_block,
+                VestingSource::Sale(sale_id)
+            )?;
+
+            // == MUTATION SAFE ==
+
+            <Joy::<T> as Currency<T::AccountId>>::transfer(
+                &sender,
+                &sale.tokens_source,
+                joy_amount,
+                ExistenceRequirement::KeepAlive
+            )?;
+
+            if account_exists {
+                AccountInfoByTokenAndAccount::<T>::mutate(token_id, &sender, |acc_data| {
+                    acc_data.add_or_update_vesting_schedule(
+                        VestingSource::Sale(sale_id),
+                        sale.get_vesting_schedule(amount),
+                        vesting_cleanup_key
+                    );
+                });
+            } else {
+                Self::deposit_to_treasury(&sender, bloat_bond);
+                Self::do_insert_new_account_for_token(
+                    token_id,
+                    &sender,
+                    AccountData::new_with_vesting_and_bond(
+                        VestingSource::Sale(sale_id),
+                        sale.get_vesting_schedule(amount),
+                        bloat_bond,
+                    )
+                );
+            }
+
+            TokenInfoById::<T>::mutate(token_id, |t| {
+                t.last_sale.as_mut().unwrap().quantity_left = sale.quantity_left.saturating_sub(amount);
+            });
+
+            Self::deposit_event(RawEvent::TokensPurchasedOnSale(token_id, sale_id, amount, sender));
+
+            Ok(())
+        }
+
+        /// Allows anyone to recover tokens that were not sold during a token sale
+        /// that has already finished (they will always be sent back to
+        /// `token_data.last_sale.tokens_source` account)
+        ///
+        /// Preconditions:
+        /// - token by `token_id` must exists
+        /// - token must be in Idle offering state
+        /// - token must have `last_sale` set
+        /// - `token_data.last_sale.quantity_left` must be > 0
+        ///
+        /// Postconditions:
+        /// - `token_data.last_sale.tokens_source` account balance is increased by
+        ///   `token_data.last_sale.quantity_left`
+        /// - `token_data.last_sale.quantity_left` is set to 0
+        #[weight = 10_000_000] // TODO: adjust weight
+        fn recover_unsold_tokens(origin, token_id: T::TokenId) -> DispatchResult {
+            ensure_signed(origin)?;
+            let token_info = Self::ensure_token_exists(token_id)?;
+            OfferingStateOf::<T>::ensure_idle_of::<T>(&token_info)?;
```

Comment:
It seems a bit akward to just keep the sale around forever, despite not being needed after a last recover, or being sold out. Wouldn't it make a simpler mental model to just require the sale to be cleaned up to recover any unsold $CRT, or ot have it done automatically when last$CRT is sold? 

Interested in your opinion here.

Reply:
Introduced the cleanup on `recover_unsold_tokens` and in case `sale.quantity_left == 0`.
Done in https://github.com/Joystream/joystream/pull/3618/commits/d8e70c04967c6c8050a12b204be1f39974e955ef